mlir/test/Dialect/Linalg/vectorize-tensor-extract.mlir

```diff
@@ -253,6 +253,54 @@ module attributes {transform.with_named_sequence} {
     transform.yield
   }
 }
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+#map1 = affine_map<(d0, d1, d2) -> (d0 + d1 + d2)>
+func.func @vectorize_nd_tensor_extract_transfer_without_outer_unit_dim(%arg0: tensor<8x128x768xf32>, %arg1 : index) -> tensor<8x1xf32> {
+  %c0 = arith.constant 0 : index
+  %0 = tensor.empty() : tensor<8x1xf32>
+  %1 = linalg.generic {
+    indexing_maps = [#map], 
+    iterator_types = ["parallel", "parallel"]
+  } outs(%0 : tensor<8x1xf32>) {
+  ^bb0(%arg5: f32):
+      %2 = linalg.index 0 : index
+      %3 = linalg.index 1 : index
+      %4 = affine.apply #map1(%arg1, %3, %arg1)
+    %extracted = tensor.extract %arg0[%2, %c0, %4] : tensor<8x128x768xf32>
```

Comment:
Isn't this a "gather load"? IIUC, this "jumps" by 768 elements on every iteration.

If it's indeed a "gather load", then `vector.transfer_read` below is not correct.

Am I missing sth here?

Reply:
You are right, this is wrong, it is being classified as a contiguous load when its not a contiguous load, looks like we need a check more conservative check here?
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L906-L909

Reply:
If this was classified a gather I believe the generated IR is correct
```
  func.func @vectorize_nd_tensor_extract_transfer_without_outer_unit_dim(%arg0: tensor<8x128x768xf32>, %arg1: index) -> tensor<8x1xf32> {
    %cst = arith.constant dense<768> : vector<1x8xindex>
    %cst_0 = arith.constant dense<128> : vector<1x8xindex>
    %c0 = arith.constant 0 : index
    %cst_1 = arith.constant dense<0.000000e+00> : vector<8x1xf32>
    %cst_2 = arith.constant dense<true> : vector<8x1xi1>
    %cst_3 = arith.constant dense<[0, 1, 2, 3, 4, 5, 6, 7]> : vector<8xindex>
    %0 = tensor.empty() : tensor<8x1xf32>
    %1 = vector.broadcast %cst_3 : vector<8xindex> to vector<1x8xindex>
    %2 = arith.addi %arg1, %arg1 : index
    %3 = vector.broadcast %2 : index to vector<1xindex>
    %4 = arith.muli %1, %cst_0 : vector<1x8xindex>
    %5 = arith.muli %4, %cst : vector<1x8xindex>
    %6 = vector.transpose %5, [1, 0] : vector<1x8xindex> to vector<8x1xindex>
    %7 = vector.broadcast %3 : vector<1xindex> to vector<8x1xindex>
    %8 = arith.addi %7, %6 : vector<8x1xindex>
    %9 = vector.gather %arg0[%c0, %c0, %c0] [%8], %cst_2, %cst_1 : tensor<8x128x768xf32>, vector<8x1xindex>, vector<8x1xi1>, vector<8x1xf32> into vector<8x1xf32>
    %10 = vector.transfer_write %9, %0[%c0, %c0] {in_bounds = [true, true]} : vector<8x1xf32>, tensor<8x1xf32>
    return %10 : tensor<8x1xf32>
  }
  ```

Reply:
@banach-space just a ping on this to see what you think.

Reply:
Hey Nirvedh, apologies for the delay - I was OOO.

> IIUC, this "jumps" by 768 elements on every iteration.

This should've been 768 * 128, right? Regardless, the overall observation holds  (i.e. this is a gather load).

> If this was classified a gather I believe the generated IR is correct

Yes, looks fine to me. 

> looks like we need a more conservative check here?
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L906-L909

This is slightly more nuanced, but should be easy to fix. First, a bit of context. 

In #102321 I've relaxed the vectoriser so that reading into a column tensor (e.g. `tensor<8x1xf32>`) is still considered as a valid candidate for "contiguous" load. Note that:

* the following comment is no longer valid (sorry about that, I will update that once we decide the right approach for this change): https://github.com/llvm/llvm-project/blob/69f3244da76586be393d1e97b01660c6f03d666c/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L893-L894
* the [test](https://github.com/llvm/llvm-project/blob/69f3244da76586be393d1e97b01660c6f03d666c/mlir/test/Dialect/Linalg/vectorize-tensor-extract.mlir#L112-L142) that I added should fail with your change - do you see why it didn't?

As for a fix, I think that what's missing is logic to identify that `%2 = linalg.index 0 : index` (which jumps by 1 on every iteration) is used for calculating a non-trailing index of `%arg0`. Look here: https://github.com/llvm/llvm-project/blob/69f3244da76586be393d1e97b01660c6f03d666c/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L897

This condition should read instead:
```cpp
  // getNonUnitLoopDim should be easy to implement
  auto nonUnitLoopDim = getNonUnitLoopDim(linalgOp);
  
  if (auto indexOp = dyn_cast<linalg::IndexOp>(defOp)) {
    foundIndexOp = (indexOp.getDim() == nonUnitLoopDim);
    return true;
  }
```

Hopefully this makes sense, but please let me know if not. This is quite convoluted stuff (mea culpa) and I really appreciate you fixing this 🙏🏻

Reply:
> the [test](https://github.com/llvm/llvm-project/blob/69f3244da76586be393d1e97b01660c6f03d666c/mlir/test/Dialect/Linalg/vectorize-tensor-extract.mlir#L112-L142) that I added should fail with your change - do you see why it didn't?

I think its becuase I had guards [here](https://github.com/llvm/llvm-project/pull/107922/files#diff-be7661e240d890a641a87f194bbb12e84a340b4a7f408865d5b89082123e24fbR1084-R1092) to not do anything for that case

> Hopefully this makes sense, but please let me know if not. This is quite convoluted stuff (mea culpa) and I really appreciate you fixing this
  
I think that does make sense to me, let me try to add the correct check.